Release notes for 2022.10 (#2483)
uxlfoundation/oneDPL: documentation/release_notes.rst

Author: It was noted that ``copy_if`` has been noted for this issue previously.  Should we leave it in the existing issues section and create a new entry for the rest, or would it be ok to keep a single entry and move it to the New Issues section?

Author: I've moved the note to the correct section.

New in 2022.10.0
================

Deprecation Notices
-------------------
The ``ONEDPL_USE_AOT_COMPILATION`` and ``ONEDPL_AOT_ARCH`` CMake options are deprecated and will be removed in a future
release. Please use the relevant compiler flags to enable this feature.

New Features
------------
- Added parallel range algorithms in ``namespace oneapi::dpl::ranges``: ``set_intersection``, ``set_union``,
  ``set_difference``, ``set_symmetric_difference``, ``includes``, ``unique``, ``unique_copy``, ``destroy``,
  ``uninitialized_fill``, ``uninitialized_move``, ``uninitialized_copy``, ``uninitialized_value_construct``,
  ``uninitialized_default_construct``, ``reverse``, ``reverse_copy``, ``swap_ranges``. These algorithms operate with
  C++20 random access ranges.
- Improved performance of ``gpu::inclusive_scan`` kernel template and added support for binary operator and type
  combinations which do not have a SYCL known identity.

> - Improved performance of ``inclusive_scan_by_segment``, ``exclusive_scan_by_segment`` and set operation algorithms when
>   using device policies: ``set_union``, ``set_difference``, ``set_intersection``, ``set_symmetric_difference``.

Reviewer: The formulation of this sentence is a bit strange.
```suggestion
- Improved performance of scan_by_segment and set operation algorithms when
  using device policies: ``inclusive_scan_by_segment``, ``exclusive_scan_by_segment``, ``set_union``, ``set_difference``, ``set_intersection``, ``set_symmetric_difference``.
```

Or alternatively, just list the algorithms out without other commentary:
``` 
- Improved performance of ``inclusive_scan_by_segment``, ``exclusive_scan_by_segment``, ``set_union``, ``set_difference``, ``set_intersection``, and ``set_symmetric_difference`` algorithms when
  using device policies.
```

Author: Separating the set operation algorithms out with the colon is odd even with the suggestion.  I've just listed the algorithms as you suggest.

  ``uninitialized_fill``, ``uninitialized_move``, ``uninitialized_copy``, ``uninitialized_value_construct``,
  ``uninitialized_default_construct``, ``reverse``, ``reverse_copy``, ``swap_ranges``. These algorithms operate with
  C++20 random access ranges.
- Improved performance of ``oneapi::dpl::experimental::kt::gpu::inclusive_scan`` and added support for binary operator
  and type combinations which do not have a SYCL known identity.
- Improved performance of ``inclusive_scan_by_segment`` and ``exclusive_scan_by_segment`` with
  device execution policies on GPUs.
- Improved performance of set operation algorithms when using device policies: ``set_union``, ``set_difference``,
  ``set_intersection``, ``set_symmetric_difference``.
- Improved performance of parallel_find_or–based algorithms on Intel® Arc™ B-series GPU devices.
- Removed requirement of GPU double support to use ``set_union``, ``set_difference``, ``set_intersection``, and
  ``set_symmetric_difference`` on Windows operating systems.
- Removed copy constructible requirements from value type for reduce and transform_reduce algorithms when used with a

>   host policy, and default constructible requirements from value type for reduce and transform_reduce when used with an execution policy.

Author: @danhoeflinger at the end of the line here is "an execution policy" a host or device policy?  Is it any policy?

Author: After offline discussion fixing this to indicate all policies are impacted.

New in 2022.10.0
================

Deprecation Notices
-------------------
The ``ONEDPL_USE_AOT_COMPILATION`` and ``ONEDPL_AOT_ARCH`` CMake options are deprecated and will be removed in a future
release. Please use the relevant compiler flags to enable this feature.

New Features
------------
- Added parallel range algorithms in ``namespace oneapi::dpl::ranges``: ``set_intersection``, ``set_union``,
  ``set_difference``, ``set_symmetric_difference``, ``includes``, ``unique``, ``unique_copy``, ``destroy``,
  ``uninitialized_fill``, ``uninitialized_move``, ``uninitialized_copy``, ``uninitialized_value_construct``,
  ``uninitialized_default_construct``, ``reverse``, ``reverse_copy``, ``swap_ranges``. These algorithms operate with
  C++20 random access ranges.

> - Improved performance of ``oneapi::dpl::experimental::kt::gpu::inclusive_scan`` and added support for binary operator

Reviewer 2: Since we've removed the `oneapi::dpl` namespace for other cases, do we just want to refer to this as "`inclusive_scan` kernel template" or keep the full namespace?

  and type combinations which do not have a SYCL known identity.
- Improved performance of ``inclusive_scan_by_segment`` and ``exclusive_scan_by_segment`` with
  device execution policies on GPUs.
- Improved performance of set operation algorithms when using device policies: ``set_union``, ``set_difference``,

>   ``set_intersection``, ``set_symmetric_difference``.

Reviewer 3: Is there a good reason not to combine these two notes?

Reviewer: Seems like they can be combined.

>   ``set_intersection``, ``set_symmetric_difference``. Removed default constructible requirements from value type of

Author: I've moved the comment.

  reduce and transform_reduce when used with an execution policy.

> - Improved performance of parallel_find_or–based algorithms on Intel® Arc™ B-series GPU devices.

Author: [83d4484](https://github.com/uxlfoundation/oneDPL/pull/2483/commits/83d448421a18728ececed3301dec96c3ba3a8d0b) lists all of the algorithms.  I think this is too much detail.  What I have in the notes now refers to the algorithms as "search algorithms" and the ``is_heap`` algorithms, which is the algorithm class from the C++ standard, to make the line more concise.

- Removed requirement of GPU double support to use ``set_union``, ``set_difference``, ``set_intersection``, and
  ``set_symmetric_difference`` on Windows operating systems.
- Removed copy constructible requirements from value type for ``reduce`` and ``transform_reduce`` algorithms when used
  with a host policy, and default constructible requirements from value type for ``reduce`` and ``transform_reduce``

>   when used with any execution policy.

Author: Done.

Fixed Issues
------------
- Fixed an issue with incorrect handling of different projections by range-based ``merge``.
- Fixed ``equal`` returning a ``false`` for empty input sequences; now it returns ``true``.
- Fixed a compilation error **SYCL kernel cannot use exceptions** occurring with libstdc++ version 10 when calling
  range-based ``adjacent_find``, ``is_sorted`` and ``is_sorted_until`` algorithms using  device policies.
- Fixed an issue with ``PSTL_USE_NONTEMPORAL_STORES`` macro having no effect.
- Fixed a bug where ``oneapi::dpl::unique`` when called with a device execution policy returned an incorrect result
  iterator.
- Fixed a bug in ``oneapi::dpl::inclusive_scan`` and all prefix sum algorithms when using device policies with differing

>   ``InputIterator`` and ``OutputIterator`` types.

Author: Done.

  ``exlusive_scan_by_segment``, and ``inclusive_scan_by_segment`` algorithms when using device policies with differing

>   ``InputIterator`` and ``OutputIterator`` types.

Reviewer: Is it differing iterator types or differing iterator value types?

> - Fixed compile errors in ``oneapi::dpl::set_union`` and ``oneapi::dpl::set_symmetric_difference`` for hetero execution

Reviewer: I think they predate #2318, but @SergeyKopienko  should weigh in.

> - Fixed compile errors in ``oneapi::dpl::set_union`` and ``oneapi::dpl::set_symmetric_difference`` for hetero execution

Reviewer 4: I am not faced with compile errors introduced in #2318. I just fixed some potentially problem places about types which describes sizes, as far as I remember.

> - Fixed error in return value types for oneapi::dpl::ranges::minmax_element and oneapi::dpl::ranges::mismatch algorithms.

Reviewer 2: ```suggestion
- Fixed error in return value types of range-based ``minmax_element`` and ``mismatch`` algorithms.
```

> - Fixed compile errors in ``set_union`` and ``set_symmetric_difference`` for hetero execution policies when the types of

Reviewer 2: I think instead of hetero we've used the term device execution policies more commonly.
```suggestion
- Fixed compile errors in ``set_union`` and ``set_symmetric_difference`` with device execution policies when the types of
```

  the second data and results data are different.

Known Issues and Limitations
----------------------------
New in This Release
^^^^^^^^^^^^^^^^^^^

> - The `set_intersection`, `set_difference`, `set_symmetric_difference`, and `set_union` algorithms with a device policy

Reviewer 2: I think this was accidentally copied here from the new issues section in the last release. We have the following note above saying it is fixed, "Removed requirement of GPU double support to use ``set_union``, ``set_difference``, ``set_intersection``, and ``set_symmetric_difference`` on Windows operating systems.".

Author: Fixed.

require GPUs with double-precision support on Windows, regardless of the value type of the input sequences.

Existing Issues
^^^^^^^^^^^^^^^
See oneDPL Guide for other `restrictions and known limitations`_.

- Range-based ``copy_if``, ``unique_copy``, ``set_union``, ``set_intersection``, ``set_difference``,
  ``set_symmetric_difference`` algorithms require the output range to have sufficient size to hold all resulting

>   elements.

Reviewer 3: Except for `copy_if`, this is a new issue.

- ``histogram`` may provide incorrect results with device policies in a program built with ``-O0`` option and the driver

>   version is 2448.13 or older.

Author: Done.

Fixed Issues
------------
- Removed requirement of GPU double precision support to use ``set_union``, ``set_difference``, ``set_intersection``,
  and ``set_symmetric_difference`` on Windows operating systems.
- Removed default-constructible requirements from the value type for ``reduce`` and ``transform_reduce`` algorithms,
  as well as copy-constructible requirements when these algorithms are used with a native ("host") policy.

> - Fixed an issue with incorrect handling of different projections by ``ranges::merge``.

Reviewer: ```suggestion
- Fixed an issue with ``ranges::merge`` when projections of the two input ranges were not the same.
```
